airflow/jobs/local_task_job.py

```diff
@@ -170,11 +174,6 @@ def heartbeat_callback(self, session=None):
                 "Terminating instance.",
                 ti.state
             )
-            if ti.state == State.FAILED and ti.task.on_failure_callback:
-                context = ti.get_template_context()
-                ti.task.on_failure_callback(context)
-            if ti.state == State.SUCCESS and ti.task.on_success_callback:
-                context = ti.get_template_context()
-                ti.task.on_success_callback(context)
             self.task_runner.terminate()
+            ti.run_finished_callback()
```

Comment:
changed the semantic to kill the task first before running callback, it would be odd to execute success callback while task is still running.